### ISSUE TITLE
Incorrect DateTime after DST change

### ISSUE THREAD

**Issue by Ferluci** [briannesbitt/Carbon#1951]
Hello,  

I encountered an issue with the following code:
```php

>>> Carbon::createFromTimestamp(1572757200, 'America/New_York')->toIso8601String();
=> "2019-11-03T01:00:00-04:00"
>>> Carbon::createFromTimestamp(1572757200+3600, 'America/New_York')->toIso8601String();
=> "2019-11-03T01:00:00-04:00"

```

Carbon version: master of github

PHP version: 7.3.8

I expected to get:

```
2019-11-03T01:00:00-05:00 like in all other languages

ruby
Time.at(1572757200).in_time_zone('America/New_York').iso8601
=> "2019-11-03T01:00:00-04:00"
irb(main):005:0> Time.at(1572757200+3600).in_time_zone('America/New_York').iso8601
=> "2019-11-03T01:00:00-05:00"

python3
>>> datetime.fromtimestamp(1572757200, pytz.timezone('America/New_York')).isoformat()
'2019-11-03T01:00:00-04:00'
>>> datetime.fromtimestamp(1572757200+3600, pytz.timezone('America/New_York')).isoformat()
'2019-11-03T01:00:00-05:00'
```

But I actually get:

```
2019-11-03T01:00:00-04:00
```

Thanks!

**Comment by kylekatarnls** [comment]
Sadly, it's an internal PHP bug, see the ton of related issues:
https://bugs.php.net/bug.php?id=51051
https://bugs.php.net/bug.php?id=61530
https://bugs.php.net/bug.php?id=62185
https://bugs.php.net/bug.php?id=62331
https://bugs.php.net/bug.php?id=63311
https://bugs.php.net/bug.php?id=63459
https://bugs.php.net/bug.php?id=65811
https://bugs.php.net/bug.php?id=66187
https://bugs.php.net/bug.php?id=67269
https://bugs.php.net/bug.php?id=68286
https://bugs.php.net/bug.php?id=69943
https://bugs.php.net/bug.php?id=72255
https://bugs.php.net/bug.php?id=72398
https://bugs.php.net/bug.php?id=72486
https://bugs.php.net/bug.php?id=74274

So by extending `DateTime`, `Carbon`actually don't have the correct timezone information during DST changes.

You would have the same problem with any PHP library or pure `DateTime`.

Working arounig this bug would require to intercept each `format()` call to check if current timestamp is in a DST period of the current timezone to override the timezone and "falsify" it.

The best I can advice is to work in UTC as long as you can. Ideally everywhere server-side, from DB to client responses. If you serve an API or webpages for example, you can pass the UTC ISO string so `"2019-11-02T21:00:00Z"` and `"2019-11-02T22:00:00Z"` for your given example, then also send the `"America/New_York"` timezone to the browser then let the browser display it in the right timezone:
```js
const date = new Date("2019-11-02T21:00:00Z");
console.log(date.toLocaleString("en-US", {timezone: "America/New_York"});
```
If the browser should display it using locale timezone, it's even easier and sending UTC dates is the best practice.

I'm sorry I can't help more, but there's nothing we can do in the Carbon code-base for that. If you're conformable with C language, you may propose a PR to PHP: https://github.com/php/php-src

**Comment by kylekatarnls** [comment]
I'm investigating deeper, I may have found a solution for this particular case.

**Comment by kylekatarnls** [comment]
In fact, this is the fix of https://bugs.php.net/bug.php?id=72338 that break this use case.

See the following:
https://3v4l.org/oWAVc

```php
$date = new DateTime('@'.(1572757200));
$date->setTimezone(new DateTimeZone('America/New_York'));
var_dump($date->format(DateTime::ATOM));
// 2019-11-03T01:00:00-04:00

$date = new DateTime('@'.(1572757200+3600));
$date->setTimezone(new DateTimeZone('America/New_York'));
var_dump($date->format(DateTime::ATOM));
// 2019-11-03T01:00:00-05:00

$date->getTimestamp();
var_dump($date->format(DateTime::ATOM));
// 2019-11-03T01:00:00-04:00
```
The simple fact to call `getTimestamp` in a `DateTime` object lose the DST shift.

I'm so desperate with this PHP bug.

**Comment by kylekatarnls** [comment]
It's even broken with `DateTimeImmutable` so `getTimestamp` modifies dates supposed to be immutable.

**Comment by kylekatarnls** [comment]
Here yet is the pure PHP version to get your case work:
```php
$date = new DateTime('@'.(1572757200));
$date->setTimezone(new DateTimeZone('America/New_York'));
var_dump($date->format(DateTime::ATOM));

$date = new DateTime('@'.(1572757200+3600));
$date->setTimezone(new DateTimeZone('America/New_York'));
var_dump($date->format(DateTime::ATOM));
```

And Carbon object with correct timezone:
```php
$date = new DateTime('@'.(1572757200));
$date->setTimezone(new DateTimeZone('America/New_York'));
var_dump(new Carbon($date->format(DateTime::ATOM)));

$date = new DateTime('@'.(1572757200+3600));
$date->setTimezone(new DateTimeZone('America/New_York'));
var_dump(new Carbon($date->format(DateTime::ATOM)));
```

But I tried to change the `createFromTimestamp` and it seems this solution may causes other problems.

**Comment by Ferluci** [comment]
@kylekatarnls 
Thank you very much for the help. You saved my life! :)

**Comment by kylekatarnls** [comment]
This will be fixed in 2.28.
#1952

**Comment by kylekatarnls** [comment]
Hi, you can try to require `"nesbot/carbon": "dev-master"` in you composer.json to test the next version and be able to run `Carbon::createFromTimestamp(1572757200 + 3600, 'America/New_York')->toIso8601String();` with the correct timezone displayed.
